app/helpers/schema_helpers.py

```diff
@@ -27,30 +27,42 @@ def wrapped_function(*args, **kwargs):
     return wrapped_function
 
 
-def get_group_instance_id(schema, answer_store, location):
+def get_group_instance_id(schema, answer_store, location, answer_instance=0):
     """Return a group instance_id if required, or None if not"""
-    if not schema.group_has_dependent_repeat_groups(location.group_id) and \
-        not schema.group_has_group_repeat_rule(location.group_id) and \
-            location.block_id not in ('relationships', 'household-relationships'):
+    if not schema.location_requires_group_instance(location):
         return None
 
-    repeat_rule = _get_group_repeat_rule(schema, location)
-    if repeat_rule:
-        group_instance_ids = []
-        for group_id in repeat_rule['group_ids']:
-            group_answer_ids = schema.get_answer_ids_for_group(group_id)
-            group_instance_ids += _get_group_instance_ids(answer_store, group_answer_ids)
+    dependent_drivers = schema.get_group_dependencies(location.group_id)
+    if dependent_drivers:
+        return _get_dependent_group_instance(schema, dependent_drivers, answer_store, location.group_instance)
 
-        return group_instance_ids[location.group_instance]
+    existing_answers = []
+    if location.group_id in schema.get_group_dependencies_group_drivers():
+        group_answer_ids = schema.get_answer_ids_for_group(location.group_id)
+        existing_answers = answer_store.filter(answer_ids=group_answer_ids, group_instance=location.group_instance)
 
-    group_answer_ids = schema.get_answer_ids_for_group(location.group_id)
-    existing_answers = answer_store.filter(answer_ids=group_answer_ids, group_instance=location.group_instance)
+    if location.block_id in schema.get_group_dependencies_block_drivers():
+        block_answer_ids = schema.get_answer_ids_for_block(location.block_id)
+        existing_answers = answer_store.filter(answer_ids=block_answer_ids, answer_instance=answer_instance)
 
     # If there are existing answers with a group_instance_id
     if existing_answers and [x for x in existing_answers if x.get('group_instance_id')]:
         return existing_answers[0]['group_instance_id']
 
-    return '-'.join([location.group_id, str(uuid4())])
+    return str(uuid4())
+
+
+def _get_dependent_group_instance(schema, dependent_drivers, answer_store, group_instance):
+    group_instance_ids = []
+    for driver_id in dependent_drivers:
+        if driver_id in schema.get_group_dependencies_group_drivers():
+            driver_answer_ids = schema.get_answer_ids_for_group(driver_id)
+            group_instance_ids += _get_group_instance_ids_for_group(answer_store, driver_answer_ids)
```

Comment:
`group_instance_ids.append(_get_group_instance_ids_for_group(answer_store, driver_answer_ids))`?

Reply:
I need extend but yes, will change that now. Otherwise i'll get a list of lists.